File: src/main/java/ar/edu/unlam/tallerweb1/delivery/ApiDispatcherServlet.java
package ar.edu.unlam.tallerweb1.delivery;

import ar.edu.unlam.tallerweb1.domain.exceptions.BusinessException;
import com.google.gson.Gson;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.io.PrintWriter;

import static ar.edu.unlam.tallerweb1.delivery.Controllers.resolve;
import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;

public class ApiDispatcherServlet extends HttpServlet {

    private Gson parser = new Gson();
    private static Logger logger = LoggerFactory.getLogger(ApiDispatcherServlet.class);
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
        doAction(request, response);
    }

    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
        doAction(request, response);
    }

    private void doAction(HttpServletRequest request, HttpServletResponse response) throws IOException {
        try{
            String json = parser.toJson(resolve(request.getRequestURI()).invoke(request));
            doResponse(response, json);
        } catch(BusinessException e){
            logger.error(e.getMessage(), e);
            response.sendError(SC_INTERNAL_SERVER_ERROR, e.getMessage());
        }catch(Exception e){
            logger.error(e.getMessage(), e);
            response.sendError(SC_INTERNAL_SERVER_ERROR);
        }
    }

    private void doResponse(HttpServletResponse response, String json) throws IOException {
        response.setContentType("application/json");
        PrintWriter printWriter = response.getWriter();
        printWriter.println(json);
        printWriter.close();
    }
}

File: src/main/java/ar/edu/unlam/tallerweb1/delivery/ApiDispatcherServlet.kt
package ar.edu.unlam.tallerweb1.delivery

import ar.edu.unlam.tallerweb1.delivery.Controllers.Companion.resolve
import ar.edu.unlam.tallerweb1.domain.exceptions.BusinessException
import com.google.gson.Gson
import org.slf4j.Logger
import org.slf4j.LoggerFactory.getLogger
import java.io.IOException
import javax.servlet.http.HttpServlet
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
import javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR

class ApiDispatcherServlet : HttpServlet() {
    private val logger: Logger = getLogger(ApiDispatcherServlet::class.java)
    private val parser = Gson()

    @Throws(IOException::class)
    override fun doGet(request: HttpServletRequest, response: HttpServletResponse) = doAction(request, response)

    @Throws(IOException::class)
    override fun doPost(request: HttpServletRequest, response: HttpServletResponse) = doAction(request, response)

    @Throws(IOException::class)
    private fun doAction(request: HttpServletRequest, response: HttpServletResponse) {
        try {
            val controller = resolve(request.requestURI)
            val json = parser.toJson(controller(request))
            doResponse(response, json)
        } catch (e: BusinessException) {
            logger.error(e.message, e)
            response.sendError(SC_INTERNAL_SERVER_ERROR, e.message)
        } catch (e: Exception) {
            logger.error(e.message, e)
            response.sendError(SC_INTERNAL_SERVER_ERROR)
        }
    }

    @Throws(IOException::class)
    private fun doResponse(response: HttpServletResponse, json: String) {
        response.contentType = "application/json"
        val printWriter = response.writer
        printWriter.println(json)
        printWriter.close()
    }
}
